hugr-passes/src/const_fold2/datalog/test.rs

```diff
@@ -0,0 +1,215 @@
+    let [in_w1, in_w2] = tlb.input_wires_arr();
+    let tail_loop = tlb.finish_with_outputs(in_w1, [in_w2, in_w1]).unwrap();
+
+    // let optype = builder.hugr().get_optype(tail_loop.node());
+    // for p in builder.hugr().node_outputs(tail_loop.node()) {
+    //     use hugr_core::ops::OpType;
+    //     println!("{:?}, {:?}", p, optype.port_kind(p));
+
+    // }
+
+    let hugr = builder.finish_hugr(&EMPTY_REG).unwrap();
+    // TODO once we can do conditionals put these wires inside `just_outputs` and
+    // we should be able to propagate their values
+    let [o_w1, o_w2, _] = tail_loop.outputs_arr();
+
+    let mut machine = Machine::new();
+    let c = machine.run_hugr(&hugr);
+    // dbg!(&machine.tail_loop_io_node);
+    // dbg!(&machine.out_wire_value);
+
+    // TODO these hould be the propagated values for now they will bt join(true,false)
+    let o_r1 = machine.read_out_wire_partial_value(&c, o_w1).unwrap();
+    // assert_eq!(o_r1, PartialValue::top());
+    let o_r2 = machine.read_out_wire_partial_value(&c, o_w2).unwrap();
+    // assert_eq!(o_r2, Value::true_val());
+    assert_eq!(
+        TailLoopTermination::Top,
+        machine.tail_loop_terminates(&c, tail_loop.node())
+    )
+}
+
+#[test]
+fn conditional() {
+    let variants = vec![type_row![], type_row![], type_row![BOOL_T]];
+    let cond_t = Type::new_sum(variants.clone());
+    let mut builder = DFGBuilder::new(FunctionType::new(Into::<TypeRow>::into(cond_t),type_row![])).unwrap();
+    let [arg_w] = builder.input_wires_arr();
+
+    let true_w = builder.add_load_value(Value::true_val());
+    let false_w = builder.add_load_value(Value::false_val());
+
+    let mut cond_builder = builder.conditional_builder((variants, arg_w), [(BOOL_T,true_w)], type_row!(BOOL_T,BOOL_T), ExtensionSet::default()).unwrap();
+    // will be unreachable
+    let case1_b = cond_builder.case_builder(0).unwrap();
+    let case1 = case1_b.finish_with_outputs([false_w,false_w]).unwrap();
+
+    let case2_b = cond_builder.case_builder(1).unwrap();
+    let [c2a] = case2_b.input_wires_arr();
+    let case2 = case2_b.finish_with_outputs([false_w,c2a]).unwrap();
+
+    let case3_b = cond_builder.case_builder(2).unwrap();
+    let [c3_1,c3_2] = case3_b.input_wires_arr();
+    let case3 = case3_b.finish_with_outputs([c3_1,false_w]).unwrap();
+
+    let cond = cond_builder.finish_sub_container().unwrap();
+
+    let [cond_o1,cond_o2] = cond.outputs_arr();
+
+    let hugr = builder.finish_hugr(&EMPTY_REG).unwrap();
+
+    let mut machine = Machine::new();
+    let arg_pv = PartialValue::variant(1, []).join(PartialValue::variant(2,[PartialValue::variant(0,[])]));
+    machine.propolutate_out_wires([(arg_w, arg_pv)]);
+    let c = machine.run_hugr(&hugr);
+
+    let cond_r1 = machine.read_out_wire_value(&c, cond_o1).unwrap();
+    assert_eq!(cond_r1, Value::false_val());
+    assert!(machine.read_out_wire_value(&c, cond_o2).is_none());
+
+    assert!(!machine.case_reachable(&c, case1.node()));
```

Comment:
@doug-q why does this pass? The predicate to the conditional node is the input to the whole Hugr here, which has Sum type with three variants; nothing tells us what argument is passed into that input??

Reply:
oh I get it - the `arg_pv`

Reply:
The argument is either variant 1 or variant 2 (containing variant 0 == false)